File: web/plugins/increase.js
import { WebPlugin } from "../engine/plugin.js";

export class IncreasePlugin extends WebPlugin {
    constructor() {
        super("増やす");
    }

    async execute(line, context) {
        const match = line.match(/^増やす\s+(.+?)\s+(.+)$/);

        if (!match) {
            throw new Error(`増やすの書式が正しくありません: ${line}`);
        }

        const [, target, amountExpression] = match;

        const amount = context.parser.parseNumber(
            amountExpression,
            context.runtime
        );

        if (context.runtime.exists(target)) {
            const current = context.parser.parseNumber(
                target,
                context.runtime
            );

            context.runtime.set(target, current + amount);
            return;
        }

        const current = context.parser.parseNumber(
            target,
            context.runtime
        );

        context.runtime.write(current + amount);
    }
}

File: pc/plugins/increase.py
import re
from core.plugin import Plugin


class IncreasePlugin(Plugin):
    def translate(self, line):
        match = re.fullmatch(
            r'増やす\s+(.+?)\s+(.+)',
            line
        )

        if not match:
            return None

        target = match.group(1).strip()
        value = match.group(2).strip()

        # 数値・式なら「結果を表示」
        if re.fullmatch(r'-?\d+(?:\.\d+)?', target):
            return f"print(({target}) + ({value}))"

        # 変数なら値を変更
        if re.fullmatch(r'[A-Za-z_][A-Za-z0-9_]*|[^\W\d]\w*', target):
            return f"{target} += {value}"

        # その他は式として計算
        return f"print(({target}) + ({value}))"